### PR TITLE
Update app manager README

### PR DESCRIPTION
**Pull request by orangejenny**
Prompted by https://github.com/dimagi/commcare-hq/issues/25718#issuecomment-546523904

### CODE REVIEW

**Review by esoergel — APPROVED** [review]
This is a great primer, thanks for beefing it up!  It will no doubt save a lot of headache down the road for those new to app_manager.

File: corehq/apps/app_manager/README.md [hunk 1]
```diff
@@ -1,21 +1,53 @@
...
+
+## Forms
+
+A "form" in HQ may refer to either a form *definition* within an application or a form *submission* containing data. App manager code typically deals with form definitions.
+
+A module contains one or more form definitions. Forms, at their most basic, are collections of questions. Forms also trigger case changes and can be configured in a variety of ways.
+
+Each form has a `unique_id` which is guaranteed unique only within the application.
+
+Forms also have an xml namespace, abbreviated `xmlns`, which is part of the form's XML definition. The xmlns uniquely identifies the form globally. Reports match form submissions to form definitions using the xmlns plus the app id, which most apps pass along to [secure_post](https://github.com/dimagi/commcare-hq/blob/5d9122ad2ba23986e6b4493eee0eab16cbcc868b/corehq/apps/receiverwrapper/views.py#L304).
```
[the remark this answers is not in the document]
Comment (orangejenny) [reply corehq/apps/app_manager/README.md]: thanks for catching that - updated in 34dbd444569

File: corehq/apps/app_manager/README.md [hunk 2]
```diff
@@ -1,21 +1,57 @@
...
-document's id is what you'll see in the URL on most app manager pages. Primary
-application documents should have `copy_of == None` and `is_released ==
-False`. When an application is saved, the field `version` is incremented.
+An application typically has many different `Application` documents: 
+one for the current/primary/canonical application, plus one for each saved build.
+The current app's id is what you'll see in the URL on most app manager pages;
+most pages will redirect to the current app if given the id of an older build.
+
+In saved builds, `copy_of` contains the primary app's id, while it's `None` for the primary app itself.
+The property `master_id` gets the id of the primary app doc, whether it's called on a saved build or on the primary app.
```
Comment (mkangia) [inline corehq/apps/app_manager/README.md]: this is in reference to Linked Applications right? I think we should specify that

File: corehq/apps/app_manager/README.md [hunk 3]
```diff
@@ -1,21 +1,57 @@
...
+
+Within code, "build" should always refer to a saved build,
+but "app" is used for both the current app and saved builds.
+The ambiguity of "app" is occasionally a source of bugs.
```
Comment (mkangia) [inline corehq/apps/app_manager/README.md]: and should be fixed where found?

**Review by mkangia — APPROVED** [review]
This was enlightening. Thanks for adding this.
Had some doubts/suggestions to add from the understanding of a new user w/o context.

File: corehq/apps/app_manager/README.md [hunk 4]
```diff
@@ -1,21 +1,57 @@
...
+## Applications (and builds)
 
-The primary application document is an instance of `Application`.  This
-document's id is what you'll see in the URL on most app manager pages. Primary
-application documents should have `copy_of == None` and `is_released ==
-False`. When an application is saved, the field `version` is incremented.
+An application typically has many different `Application` documents: 
+one for the current/primary/canonical application, plus one for each saved build.
+The current app's id is what you'll see in the URL on most app manager pages;
+most pages will redirect to the current app if given the id of an older build.
```
[the remark this answers is not in the document]
Comment (orangejenny) [reply corehq/apps/app_manager/README.md]: Saved builds are pretty much read-only, so it doesn't make sense to edit them in app manager.

[the remark this answers is not in the document]
Comment (orangejenny) [reply corehq/apps/app_manager/README.md]: (adding note to docs)

File: corehq/apps/app_manager/README.md (hunk 2 above)
Comment (orangejenny) [reply corehq/apps/app_manager/README.md]: This is not related to linked apps, though I agree it's confusing that they booth use "master" - this is why linked apps now use `upstream_app_id`.

File: corehq/apps/app_manager/README.md (hunk 3 above)
Comment (orangejenny) [reply corehq/apps/app_manager/README.md]: Probably, although this ambiguity is all over the place and it's often difficult to decide how to rename - like, `get_app` gets both apps and builds, should we call it `get_app` or `get_app_or_build`, or have both a `get_app` and a `get_build`?

File: corehq/apps/app_manager/README.md [hunk 5]
```diff
@@ -1,21 +1,57 @@
...
 
-When a build is starred, this is called "releasing" the build.  The parameter
-`is_released` will be set to True on the build document.
+### Exceptions
+Linked apps use similar workflows to app copy for creating and pulling. See [docs](https://github.com/dimagi/commcare-hq/tree/master/corehq/apps/linked_domain#linked-applications) for more detail on how they handle form unique ids and xmlnses.
 
-You might also run in to "remote" applications and applications copied to be
-"published on the exchange", but those are quite infrequent.
+Shadow forms are a variation of advanced forms that "shadow" another form's XML but can have their own settings and actions. Because they don't have their own XML, shadow forms do not have an xmlns but instead inherit their parent's xmlns.
```
[the remark this answers is not in the document]
Comment (orangejenny) [reply corehq/apps/app_manager/README.md]: Updated docs - they show up in reports as though they were submissions of the parent form.

File: corehq/apps/app_manager/README.md (hunk 3 above)
Comment (mkangia) [reply corehq/apps/app_manager/README.md]: i think that method let's the user not worry about what they need and just give an `id` so `get_app_or_build` sounds right.
However i think this line was talking about variables used but better method names would be good too.

File: corehq/apps/app_manager/README.md (hunk 2 above)
Comment (mkangia) [reply corehq/apps/app_manager/README.md]: hmm, i am getting confused between `copy_of` and `master_id`. It's not clear why are these two different. "primary app's id" and "id of the primary app doc" appear to be the same.

File: corehq/apps/app_manager/README.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (orangejenny) [reply corehq/apps/app_manager/README.md]: Hmm, I don't agree that anything unexpected should raise an error. If we can recover gracefully, we should.

Form designer includes a user-facing message: https://github.com/dimagi/commcare-hq/blob/3c0d08989efe82b412c20e2400cc646ba6f3f40f/corehq/apps/app_manager/views/formdesigner.py#L109-L113

I'd be open to adding something like that in the other views that do this kind of redirect, but not to throwing a harder error.

File: corehq/apps/app_manager/README.md (hunk 2 above)
Comment (orangejenny) [reply corehq/apps/app_manager/README.md]: `_id` is for primary apps, `copy_of` is for builds, and `master_id` is for code that has to handle both primary apps and builds

File: corehq/apps/app_manager/README.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (mkangia) [reply corehq/apps/app_manager/README.md]: Cool, My intent was that we should let the user know that we did something behind the scene.

**Pull request merged**